typing/typecore.ml

```diff
@@ -2697,11 +2705,23 @@ and type_expect_
   | Pexp_match(sarg, caselist) ->
       begin_def ();
       let arg = type_exp env sarg in
-      end_def ();
+       end_def ();
       if maybe_expansive arg then lower_contravariant env arg.exp_type;
       generalize arg.exp_type;
       let cases, partial =
         type_cases Computation env arg.exp_type ty_expected true loc caselist in
+        let contains_exception_pattern pat =
+        exists_general_pattern { f = fun (type k) (p : k general_pattern) ->
+          match p.pat_desc with
+          | Tpat_exception _ -> true
+          | _ -> false } pat
```

Comment:
This is a quite self-contained function that is used in two different places. It is often better in this situation to make the function a toplevel helper function shared between the two call sites.